File: src/decoders/notebook-absence.ts
import type { NotebookAbsence } from "~/models";
import { decodePronoteDate } from "./pronote-date";

export const decodeNotebookAbsence = (absence: any): NotebookAbsence => {
  const [hoursMissed, minutesMissed] = (absence.NbrHeures as string).split("h").map(Number);
  const isReasonUnknown = absence.estMotifNonEncoreConnu;

  return {
    id: absence.name,
    startDate: decodePronoteDate(absence.dateDebut.V),
    endDate: decodePronoteDate(absence.dateFin.V),
    justified: absence.justifie,
    opened: absence.ouverte,
    hoursMissed, minutesMissed,
    daysMissed: absence.NbrJours,
    shouldParentsJustify: absence.aJustifierParParents,
    administrativelyFixed: absence.reglee,
    isReasonUnknown,
    reason: !isReasonUnknown && absence.listeMotifs.V[0].L
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/notebook-absence.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.NotebookAbsence
import kotlinx.serialization.json.*

fun decodeNotebookAbsence(absence: JsonObject): NotebookAbsence {
  val (hoursMissed, minutesMissed) =
      absence["NbrHeures"]!!.jsonPrimitive.content.split('h').map { it.toInt() }
  val isReasonUnknown = absence["estMotifNonEncoreConnu"]!!.jsonPrimitive.boolean

  return NotebookAbsence(
      id = absence["N"]!!.jsonPrimitive.content,
      startDate = decodePronoteDate(absence["dateDebut"]!!.jsonObject["V"]!!.jsonPrimitive.content),
      endDate = decodePronoteDate(absence["dateFin"]!!.jsonObject["V"]!!.jsonPrimitive.content),
      justified = absence["justifie"]!!.jsonPrimitive.boolean,
      opened = absence["ouverte"]!!.jsonPrimitive.boolean,
      hoursMissed = hoursMissed,
      minutesMissed = minutesMissed,
      daysMissed = absence["NbrJours"]!!.jsonPrimitive.double,
      shouldParentsJustify = absence["aJustifierParParents"]!!.jsonPrimitive.boolean,
      administrativelyFixed = absence["reglee"]!!.jsonPrimitive.boolean,
      isReasonUnknown = isReasonUnknown,
      reason =
          if (!isReasonUnknown)
              absence["listeMotifs"]!!
                  .jsonObject["V"]!!
                  .jsonArray[0]
                  .jsonObject["L"]!!
                  .jsonPrimitive
                  .content
          else null)
}
